Move Router filter to V2 api

rshriram opened the pull request:
*title*: V2 support for router filter

*Description*: Enable parsing router filter proto
*Risk Level*: Medium
*Testing*: unit tests, existing JSON tests

rshriram commented:
@zuercher FYI. this might change your upcoming PR

rshriram commented:
@junr03 fyi

zuercher commented:
@rshriram the conflict will be in `source/server/config/http/router.cc`. In my change I translate the JSON object into an envoy::api::v2::filter::http::Router and pass that to Router::FilterConfig. Let it extract the values it needs.

mattklein123 commented on source/server/config/http/fault.cc:
```diff
@@ -5,8 +5,6 @@
 #include "common/config/filter_json.h"
 #include "common/http/filter/fault_filter.h"
 
-#include "api/filter/http/fault.pb.h"
```
?

rshriram replied:
This was an unnecessary include. The header was already including it.

mattklein123 commented on source/server/config/http/router.cc:
```diff
@@ -12,23 +12,42 @@ namespace Envoy {
+                                                            const std::string& stat_prefix,
+                                                            FactoryContext& context) {
+  json_config.validateSchema(Json::Schema::ROUTER_HTTP_FILTER_SCHEMA);
+  return createRouterFilter(json_config.getBoolean("dynamic_stats", true),
```
Per @zuercher can we convert to proto and pass? This will be easier in the future as we add more config.

zuercher replied:
You can look at my change here: https://github.com/envoyproxy/envoy/compare/master...turbinelabs:implement-upstream-log#diff-ac294912df489e3b4f8094bdb1e91918

I can also convert this when I merge, so I added a translate function in FilterJson, a schema in config_schema.cc, tests, etc... It's a whole thing.

mattklein123 replied:
@zuercher maybe you and @rshriram can coordinate over slack briefly? Seem like he could just copy part of your PR or could split off part of your PR. Either way.

mattklein123 commented on source/server/config/http/router.cc:
```diff
@@ -12,23 +12,42 @@ namespace Envoy {
+    const Protobuf::Message& config, const std::string& stat_prefix, FactoryContext& context) {
+  const envoy::api::v2::filter::http::Router& router =
+      dynamic_cast<const envoy::api::v2::filter::http::Router&>(config);
+  bool dynamic_stats = true; // defaults to true
```
Can we use PROTOBUF_GET_WRAPPED_OR_DEFAULT here?

mattklein123 commented on source/server/config/http/router.cc:
```diff
@@ -12,23 +12,42 @@ namespace Envoy {
+  if (router.has_dynamic_stats()) {
+    dynamic_stats = router.dynamic_stats().value();
+  }
+  bool start_child_span = router.start_child_span(); // defaults to false
```
nit: comment probably not needed.

mattklein123 commented on source/server/config/http/router.cc:
```diff
@@ -5,30 +5,47 @@
+    const Protobuf::Message& config, const std::string& stat_prefix, FactoryContext& context) {
+  const envoy::api::v2::filter::http::Router& router =
+      dynamic_cast<const envoy::api::v2::filter::http::Router&>(config);
+  bool dynamic_stats = PROTOBUF_GET_WRAPPED_OR_DEFAULT(router, dynamic_stats, true);
```
sorry, more nit: const here and next line, or just do inline in call.

mattklein123 commented:
@rshriram per offline convo can you also fix the tests? I think you could probably do something like:
```
envoy::api::v2::filter::http::Router config = dynamic_cast<...>(factory.createEmptyProto());
...
HttpFilterFactoryCb cb = factory.createFilterFactoryFromProto(config, "stats", context);
```

But we should also probably have HTTP connection manager and listener manager tests that v2 filters actually load. Will defer to @htuch on best way of testing this.

rshriram commented:
@mattklein123 / @zuercher / @htuch PTAL. Sorry for the delay.
I have added tests for the createEmptyProtoConfig function, and I have also added those functions to the other filters I added (fault/mongo). I am not really sure if its testing the entire flow though.
Looking at the code, translateToFactoryConfig is where the action happens. But I can't see any test that is invoking it.

Is there an example code for testing this code path, that I can follow?

mattklein123 commented:
> Is there an example code for testing this code path, that I can follow?

In general the flow is covered by the v2 integration tests we have. Right now the HTTP connection manager is loaded via pure v2. Currently, all HTTP filters are loaded using v1_deprecated. What I would do is search for v1_deprecated in the v2 integration tests, and remove it as you convert.

For this PR, sorry to be annoying, but I think a better test beyond what you have is what I was saying above (https://github.com/envoyproxy/envoy/pull/1982#issuecomment-341292916). Essentially, create an empty proto and dynamic cast copy it to a local var, which will check that it is correct, then fill it. I would do this for each of the filters you have converted so far.

mattklein123 commented:
@rshriram just merged @zuercher PR so this is going to need a merge along with the testing stuff described above.

mattklein123 commented on source/server/config/network/mongo_proxy.h:
```diff
@@ -23,6 +23,10 @@ class MongoProxyFilterConfigFactory : public NamedNetworkFilterConfigFactory {
                                                       FactoryContext& context) override;
 
+  ProtobufTypes::MessagePtr createEmptyConfigProto() override {
+    return ProtobufTypes::MessagePtr{new envoy::api::v2::filter::network::MongoProxy()};
```
Is this covered? If not can we do the strategy that I said? (Assign the return of this to a config variable than fill it manually)

rshriram replied:
done
